ssl/statem/statem_lib.c

```diff
@@ -2415,3 +2415,162 @@ int tls13_restore_handshake_digest_for_pha(SSL *s)
     }
     return 1;
 }
+
+#if !defined(OPENSSL_NO_COMP)
+
+/* Intended to be used after WPACKET_reserve_bytes() */
+static int wpacket_advance_write(WPACKET *pkt, size_t len)
+{
+    if (pkt->maxsize - pkt->written < len)
+        return 0;
+    if (pkt->staticbuf == NULL && (pkt->buf->length - pkt->written < len))
+        return 0;
+
+    pkt->written += len;
+    pkt->curr += len;
+    return 1;
+}
+
+int tls13_construct_compressed_certificate(SSL *s, WPACKET *pkt, CERT_PKEY *cpk,
+                                           uint8_t *context, size_t context_len)
+{
+    WPACKET tmppkt;
+    BUF_MEM *buf = NULL;
+    int ret = 0;
+    size_t length;
+    size_t max_length;
+    COMP_METHOD *method;
+    COMP_CTX* comp = NULL;
+    int comp_len;
+
+    if (s->ext.compress_certificate_rx == TLSEXT_comp_cert_none
+            || (buf = BUF_MEM_new()) == NULL
+            || !WPACKET_init(&tmppkt, buf)) {
+        goto err;
+    }
+
+    /* Use the |tmppkt| for the to-be-compressed data */
+    if (context == NULL) {
+        /* no context available, add 0-length context */
+        if (!WPACKET_put_bytes_u8(&tmppkt, 0))
+            goto err;
+    } else if (!WPACKET_sub_memcpy_u8(&tmppkt, context, context_len))
+        goto err;
+
+    if (!ssl3_output_cert_chain(s, &tmppkt, cpk)) {
+        /* SSLfatal() already called */
+        goto out;
+    }
+
+    /* continue with the real |pkt| */
+    if (!WPACKET_put_bytes_u16(pkt, s->ext.compress_certificate_rx)
+            || !WPACKET_get_total_written(&tmppkt, &length)
+            || !WPACKET_put_bytes_u24(pkt, length))
+        goto err;
+
+    /*
+     * Uncompressibility expansion:
+     * Brotli: per RFC7932: N + 5 + 3 * (N >> 16)
+     * ZLIB: N + 11 + 5 * (N >> 14)
+     * ZSTD: N + 4 + 14  \+ 3 * (N >> 17) + 4
+     */
+    switch (s->ext.compress_certificate_rx) {
+# ifdef ZLIB
+    case TLSEXT_comp_cert_zlib:
+        max_length = length + 11 + 5 * (length >> 14);
+        method = COMP_zlib();
+        break;
+# endif
+# ifdef BROTLI
+    case TLSEXT_comp_cert_brotli:
+        max_length = length + 5 + 3 * (length >> 16);
+        method = COMP_brotli_oneshot();
+        break;
+# endif
+# ifdef ZSTD
+    case TLSEXT_comp_cert_zstd:
+        max_length = length + 22 + 3 * (length >> 17);
+        method = COMP_zstd_oneshot();
+        break;
+# endif
+    default:
+        goto err;
+    }
+
+    if (!WPACKET_reserve_bytes(pkt, max_length+3, NULL)
+            || (comp = COMP_CTX_new(method)) == NULL)
+        goto err;
+
+    comp_len = COMP_compress_block(comp, WPACKET_get_curr(pkt)+3, max_length,
```

Comment:
So we compress every time? Is there no opportunity to cache this?

Reply:
Callback API is added in later commit to allow a user to cache it.

Reply:
Also, due to the context value, only the server-side would be able to cache.